mijia/src/bluetooth/events.rs

```diff
@@ -0,0 +1,205 @@
+use dbus::arg::prop_cast;
+use dbus::message::{MatchRule, SignalArgs};
+use dbus::nonblock::stdintf::org_freedesktop_dbus::PropertiesPropertiesChanged;
+use dbus::Message;
+
+use super::{AdapterId, CharacteristicId, DeviceId};
+
+/// An event relating to a Bluetooth device or adapter.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum BluetoothEvent {
+    /// An event related to a Bluetooth adapter.
+    Adapter {
+        /// The ID of the Bluetooth adapter in question.
+        id: AdapterId,
+        /// Details of the specific event.
+        event: AdapterEvent,
+    },
+    /// An event related to a Bluetooth device.
+    Device {
+        /// The ID of the Bluetooth device in question.
+        id: DeviceId,
+        /// Details of the specific event.
+        event: DeviceEvent,
+    },
+    /// An event related to a GATT characteristic of a Bluetooth device.
+    Characteristic {
+        /// The ID of the GATT characteristic in question.
+        id: CharacteristicId,
+        /// Details of the specific event.
+        event: CharacteristicEvent,
+    },
+}
+
+/// Details of an event related to a Bluetooth adapter.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum AdapterEvent {
+    /// The adapter has been powered on or off.
+    Powered { powered: bool },
+    /// The adapter has started or stopped scanning for devices.
+    Discovering { discovering: bool },
+}
+
+/// Details of an event related to a Bluetooth device.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DeviceEvent {
+    /// The device has connected or disconnected.
+    Connected { connected: bool },
+    /// A new value is available for the RSSI of the device.
+    RSSI { rssi: i16 },
+}
+
+/// Details of an event related to a GATT characteristic.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CharacteristicEvent {
+    /// A new value of the characteristic has been received. This may be from a notification.
+    Value { value: Vec<u8> },
+}
+
+impl BluetoothEvent {
+    /// Return a `MatchRule` which will match all D-Bus messages which represent Bluetooth events.
+    pub(crate) fn match_rule() -> MatchRule<'static> {
+        // BusName validation just checks that the length and format is valid, so it should never
+        // fail for a constant that we know is valid.
+        let bus_name = "org.bluez".into();
```

Comment:
comment seems stale

Reply:
It's still valid. `into()` will panic if the string is invalid, but that won't happen for this constant string.